src/AllowList.php

```diff
@@ -5,78 +5,82 @@
 namespace Laminas\Filter;
 
 use Laminas\Stdlib\ArrayUtils;
-use Traversable;
 
+use function array_values;
 use function in_array;
 use function is_array;
 
 /**
  * @psalm-type Options = array{
  *     strict?: bool,
- *     list?: array,
- *     ...
+ *     list?: iterable<array-key, mixed>,
  * }
- * @extends AbstractFilter<Options>
- * @final
  */
-class AllowList extends AbstractFilter
+final class AllowList implements FilterInterface
 {
-    /** @var bool */
-    protected $strict = false;
+    private bool $strict = false;
+    /** @var list<mixed> */
+    private array $list = [];
 
-    /** @var array */
-    protected $list = [];
+    /**
+     * @param Options $options
+     */
+    public function __construct(array $options = [])
+    {
+        $this->setOptions($options);
+    }
 
     /**
-     * @param null|array|Traversable $options
+     * @param Options $options
+     * @return $this
      */
-    public function __construct($options = null)
+    public function setOptions(array $options = []): self
```

Comment:
I'd say lets remove `setOptions` at all.
Filters should be immutable once instantiated, there is *usually* no need to modify options at runtime and thus this method should be removed.

All of the code here can be handled inside of `__construct`.

Reply:
I'm in complete agreement here. Dropping `setOptions` would be ideal, however, I'll have to check this out - mainly, I'm not sure how the `FilterPluginManager` instantiates filters - IIRC, it's a case of

```php
$filter = new SomeFilter(); // Or build from factory
$filter->setOptions($options);
```

I could be wrong of course!

Reply:
You can change the filter plugin manager as well.
AFAIK, all filters which do not have dedicated options are usually depending on `InvokableFactory` which only does:

```php
new $filter($options);
```

Reply:
TBH, the reason I started doing these improvements in the first place was to fit them all in to the next major whilst providing SMv4 support.

I'll amend the patch to kill off all runtime mutation of filters and then deal with the plugin manager later 👍